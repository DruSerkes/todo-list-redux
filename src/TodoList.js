import React from 'react';
import Todo from './Todo';
import NewTodoForm from './NewTodoForm';
import EditTodoForm from './EditTodoForm';
import { v4 as uuid } from 'uuid';
import { useSelector, useDispatch } from 'react-redux';

/**
 * TodoList - this component renders the NewTodoForm component 
 * and a list of Todo components. 
 */

const TodoList = () => {
	const todos = useSelector((state) => state.todos);
	const dispatch = useDispatch();

	const addTodo = ({ text }) => {
		if (!text) return;
		const newTodo = { text, id: uuid(), complete: false, editing: false };
		const updatedTodos = [ ...todos, newTodo ];
		updateLocalStorage(updatedTodos);
		dispatch({ type: 'ADD_TODO', payload: newTodo });
	};

	const updateLocalStorage = (updatedTodos) => {
		localStorage.setItem('todos', JSON.stringify(updatedTodos));
	};

	const removeTodo = (id) => {
		const storedTodos = JSON.parse(localStorage.getItem('todos'));
		const updatedTodos = storedTodos.filter((todo) => todo.id !== id);
		updateLocalStorage(updatedTodos);
		dispatch({ type: 'REMOVE_TODO', id });
	};

	const toggleComplete = (id) => {
		const todo = todos.find((todo) => todo.id === id);
		todo.complete = !todo.complete;
		const updatedTodos = [ ...todos ];
		updateLocalStorage(updatedTodos);
		dispatch({ type: 'TOGGLE_COMPLETE', payload: updatedTodos });
	};

	const toggleEditing = (id) => {
		const todo = todos.find((todo) => todo.id === id);
		todo.editing = !todo.editing;
		const updatedTodos = [ ...todos ];
		updateLocalStorage(updatedTodos);
		dispatch({ type: 'TOGGLE_EDITING', payload: updatedTodos });
	};

	const editTodo = ({ id, text }) => {
		if (!text) return;
		const todo = todos.find((todo) => todo.id === id);
		todo.editing = !todo.editing;
		todo.text = text;
		const updatedTodos = [ ...todos ];
		updateLocalStorage(updatedTodos);
		dispatch({ type: 'EDIT_TODOS', payload: updatedTodos });
	};

	const checkIfEditing = (todo) => {
		return todo.editing ? (
			<EditTodoForm text={todo.text} id={todo.id} editTodo={editTodo} key={todo.id} />
		) : (
			<Todo
				text={todo.text}
				key={todo.id}
				id={todo.id}
				complete={todo.complete}
				removeTodo={removeTodo}
				toggleEditing={toggleEditing}
				toggleComplete={toggleComplete}
			/>
		);
	};

	return (
		<div className="TodoList">
			<h1 className="TodoList-Header">My Todo List</h1>
			<NewTodoForm addTodo={addTodo} />
			<div className="TodoList-Container">{todos.map((todo) => checkIfEditing(todo))}</div>
		</div>
	);
};

export default TodoList;
