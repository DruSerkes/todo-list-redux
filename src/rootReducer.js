const todos = JSON.parse(localStorage.getItem('todos')) || [];
const INITIAL_STATE = { todos };

const rootReducer = (state = INITIAL_STATE, action) => {
	switch (action.type) {
		case 'ADD_TODO':
			return { ...state, todos: [ ...state.todos, action.payload ] };
		case 'REMOVE_TODO':
			return { ...state, todos: state.todos.filter((todo) => todo.id !== action.id) };
		case 'TOGGLE_COMPLETE':
			return { ...state, todos: action.payload };
		case 'TOGGLE_EDITING':
			return { ...state, todos: action.payload };
		case 'EDIT_TODOS':
			return { ...state, todos: action.payload };
		case 'REMOVE_ALL':
			return { ...state, todos: [] };
		default:
			return state;
	}
};

export default rootReducer;
