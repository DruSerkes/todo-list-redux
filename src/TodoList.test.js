import React from 'react';
import { render, fireEvent, cleanup } from '@testing-library/react';
import TodoList from './TodoList';
import { Provider } from 'react-redux';
import { createStore } from 'redux';
import rootReducer from './rootReducer';

const store = createStore(rootReducer);

// Smoke test
it('renders without crashing', () => {
	render(
		<Provider store={store}>
			<TodoList />
		</Provider>
	);
});

// Snapshot test
it('matches snapshot', () => {
	const { asFragment } = render(
		<Provider store={store}>
			<TodoList />
		</Provider>
	);
	expect(asFragment()).toMatchSnapshot();
});

describe('TodoList tests', () => {
	afterAll(() => {
		cleanup();
		localStorage.clear();
	});

	it('should add new todo', () => {
		const { queryByText, getByPlaceholderText } = render(
			<Provider store={store}>
				<TodoList />
			</Provider>
		);
		expect(queryByText('breakfast')).not.toBeInTheDocument();

		const input = getByPlaceholderText('Breakfast...', { exact: false });
		const btn = queryByText('Add');
		fireEvent.change(input, { target: { value: 'breakfast' } });
		fireEvent.click(btn);
		expect(queryByText('breakfast')).toBeInTheDocument();
	});

	it('should remove a todo', () => {
		const { queryByText, getByPlaceholderText } = render(
			<Provider store={store}>
				<TodoList />
			</Provider>
		);

		expect(queryByText('breakfast')).toBeInTheDocument();
		const removeBtn = queryByText('â…¹');
		fireEvent.click(removeBtn);
		expect(queryByText('breakfast')).not.toBeInTheDocument();
	});
});
