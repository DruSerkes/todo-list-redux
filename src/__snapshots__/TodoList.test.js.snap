// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="TodoList"
  >
    <h1
      class="TodoList-Header"
    >
      My Todo List
    </h1>
    <div
      class="NewTodoForm"
    >
      <form>
        <label
          for="text"
        >
          Add To do: 
        </label>
        <input
          id="text"
          name="text"
          placeholder="Eat Breakfast..."
          type="text"
          value=""
        />
        <button
          class="NewTodoForm-Add"
        >
          Add
        </button>
      </form>
    </div>
    <div
      class="TodoList-Container"
    />
  </div>
</DocumentFragment>
`;
